Checkpointing support for transformer type models

zhenghh04 opened the pull request:
In this PR, we addressed the issue that people have to manually input layer parameters and optimization groups in the checkpointing. https://github.com/argonne-lcf/dlio_benchmark/issues/248

* People are allowed to input just "vocab_size, hidden_size, ffn_hidden_size, num_layers", the layer parameters and optimization groups are calculated internally. 
* We also added option for specifying zero_stage. The default value is 0, where no zero is used. 
* We added several YAML configuration files for Llama transformer models. 
* We allow people to specify the datatype for writing data. 
* We added checkpoint only mode
* We added support for checkpoint recovery tests

zhenghh04 commented:
@hariharan-devarajan This PR is ready for review.

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
+            self.optimization_groups_predefined = True
+            return self.args.optimization_groups
+        else:
+            if self.args.num_layers <= 0:
+                return []
+            if self.args.zero_stage > 0:
+                # zero stage 1, 2, 3
+                num_p = self.get_num_parameters() // self.args.comm_size
+            else:
+                # if zero is not used. Only the first data parallel instance will save the optimizer states
+                num_p = self.get_num_parameters() // self.mp
```
In this code, why zero_stage=0 can't determine the first data parallel instance will save the optimizer states. We dont need -1 right? Or does zero_stage has more meaning?

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -86,22 +240,55 @@ def get_layer_index(self, rank, tensor_parallelism, pipeline_parallelism, total_
             start_layer = left_layers * max_layer_per_pipeline + (pipeline_rank - left_layers) * (min_layers_per_pipeline)
             end_layer = start_layer + num_layers_per_pipeline - 1
         return start_layer, end_layer
+        '''
```
I am fine with your removing the comment as we have git to check what the code was.

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
+        self.MPI = DLIOMPI.get_instance()
+        self.comm = self.MPI.comm()
+        # define parallelism
+        self.pp = self.args.pipeline_parallelism
+        self.tp = self.args.tensor_parallelism
+        self.mp = self.tp*self.pp
+        self.dp = self.args.comm_size//self.mp
+        self.pp_rank = (self.args.my_rank // self.tp) % self.pp
+        self.tp_rank = self.args.my_rank % self.tp
+        self.dp_rank = self.args.my_rank // (self.pp*self.tp)
+        self.mp_rank = self.args.my_rank%self.mp
```
variable like pp, tp, mp, and dp are too abbreviated. I know what they mean but they reduce code-readability. I would recommend spelling them out. You do this throughout the code. I will try to mark them as much as I can but please make sure we dont use abbreviated variable names.

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
+        self.layer_parameters_predefined = False
+        self.checkpoint_storage.create_namespace(exist_ok=True)
+        self.rank_to_checkpoint = self.args.my_rank
+        self.num_p = self.get_num_parameters()
```
non abbreviate

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
+        self.rank_to_checkpoint = self.args.my_rank
+        self.num_p = self.get_num_parameters()
+        self.checkpoint_size = 0.0
+        ss = 0.0
```
non abbreviate

hariharan-devarajan replied:
What is ss. This was not fixed. I would recommend to just comment if you have fixed. So that I can review and resolve if it looks fine.

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
+                    self.layer_state[str(layer_index)] = layer_state  
+            else:
+                self.layer_state = dict()
+                ss = 0.0
```
non abbreviate

hariharan-devarajan replied:
Again what is ss.

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
+            if self.args.my_rank == 0:
+                logging.info(f"{utcnow()} Model state defined")
+
+        ss = self.comm.allreduce(ss)/1024./1024./1024.
```
non abbreviate

hariharan-devarajan replied:
fix what is ss.

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
+                logging.info(f"{utcnow()} Model state defined")
+
+        ss = self.comm.allreduce(ss)/1024./1024./1024.
+        opt = self.comm.allreduce(self.checkpoint_size)/1024./1024./1024.
```
non abbreviate. I this optimization? or optimization_size?

hariharan-devarajan replied:
again not fixed.

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
+        if l <= 0:
+            return 0
+        embedding = voc*h
+        input_norm = h
+        qkv = 3*h*h
+        dense = h*h
+        layer_norm = h
+        mlp_h_to_4h = ffn*2*h # the factor of 2 is because of gated linear unit                                                                           
+        mlp_4h_to_h = ffn*h
+        weight = h
+        lm_head = embedding
```
fix abbreviation.

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
 
-    def get_layer_index(self, rank, tensor_parallelism, pipeline_parallelism, total_layers):
+    def get_num_parameters(self):
+        h, l, ffn, voc = self.args.hidden_size, self.args.num_layers, self.args.ffn_hidden_size, self.args.vocab_size
```
fix abbreviated variables.

hariharan-devarajan replied:
what is h, l, ffn, voc. This was not fixed.

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
+                sharding_factor = 1
+            else:
+                sharding_factor = self.dp
+            h, l, ffn, voc = self.args.hidden_size, self.args.num_layers, self.args.ffn_hidden_size, self.args.vocab_size
```
fix abbreviated variables.

hariharan-devarajan replied:
again this was not fixed, h, l, ffn, voc

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -21,53 +21,207 @@
+        return layer_state, size
+
+    def get_optimization_groups(self):
+        h, l, ffn, voc = self.args.hidden_size, self.args.num_layers, self.args.ffn_hidden_size, self.args.vocab_size
```
fix abbreviated variables.

hariharan-devarajan replied:
again this was not fixed. h, l, ffn, voc

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -86,22 +240,55 @@ def get_layer_index(self, rank, tensor_parallelism, pipeline_parallelism, total_
+        We assume layer 0 is always on rank 0, and l+1 and l+2 are on the last rank.                                                                      
+        '''
+        pipeline_rank = self.pp_rank
+        nl = self.args.num_layers//self.pp
```
fix abbreviated variables.

hariharan-devarajan commented on dlio_benchmark/checkpointing/pytorch_checkpointing.py:
```diff
@@ -42,14 +59,16 @@ def __init__(self):
-    def get_tensor(self, size):
-        return torch.randint(high=1, size=(size,), dtype=torch.int8)
+    def get_tensor(self, size, datatype="int8"):
+        return torch.ones(size, dtype=get_torch_datatype(datatype))
```
I had tried torch.ones, the tensor representation in memory was smaller as PyTorch was doing some optimization. that's why I used randint. Can you make sure the I/O and memory representation remain same?

hariharan-devarajan replied:
sounds good, can you fix the size variable to be length?

hariharan-devarajan commented on dlio_benchmark/checkpointing/pytorch_checkpointing.py:
```diff
@@ -42,14 +59,16 @@ def __init__(self):
     @dlp.log
-    def get_tensor(self, size):
-        return torch.randint(high=1, size=(size,), dtype=torch.int8)
+    def get_tensor(self, size, datatype="int8"):
```
The API is different from what I had. What I mean is size is inclusive of type of data. Based on what you want to do, you can do array_length? or length instead of size.

hariharan-devarajan replied:
can you fix the size variable to be length?

hariharan-devarajan commented on dlio_benchmark/checkpointing/tf_checkpointing.py:
```diff
@@ -42,11 +58,11 @@ def __init__(self):
     @dlp.log
-    def get_tensor(self, size):
-        return tf.random.uniform((int(size / 4),), maxval=100, dtype=tf.dtypes.int32)
+    def get_tensor(self, size, datatype="int8"):
```
The API is different from what I had. What I mean is size is inclusive of type of data. Based on what you want to do, you can do array_length? or length instead of size.

hariharan-devarajan replied:
change size to length

hariharan-devarajan commented on dlio_benchmark/checkpointing/tf_checkpointing.py:
```diff
@@ -42,11 +58,11 @@ def __init__(self):
-    def get_tensor(self, size):
-        return tf.random.uniform((int(size / 4),), maxval=100, dtype=tf.dtypes.int32)
+    def get_tensor(self, size, datatype="int8"):
+        return tf.ones((size), dtype=get_tf_datatype(datatype))
```
Same comment as PyTorch is to make sure tensor representation has same footprint.

hariharan-devarajan commented on dlio_benchmark/configs/workload/llama_1t.yaml:
```diff
@@ -0,0 +1,47 @@
+checkpoint:
+  checkpoint_folder: checkpoints/llama_405b
+  steps_between_checkpoints: 1
+  model_datatype: fp16
+  optimizer_datatype: fp32
```
I feel these should be in model.

hariharan-devarajan commented on dlio_benchmark/configs/workload/llama_405b.yaml:
```diff
@@ -0,0 +1,46 @@
+checkpoint:
+  checkpoint_folder: checkpoints/llama_405b
+  steps_between_checkpoints: 1
+  model_datatype: fp16
+  optimizer_datatype: fp32
```
I feel these should be in model.

hariharan-devarajan commented on dlio_benchmark/configs/workload/llama_70b.yaml:
```diff
@@ -0,0 +1,46 @@
+checkpoint:
+  checkpoint_folder: checkpoints/llama_70b
+  steps_between_checkpoints: 1
+  model_datatype: fp16
+  optimizer_datatype: fp32
```
I feel these should be in model. and you missed adding a new line at end of file here.

hariharan-devarajan replied:
These were not fixed.

hariharan-devarajan commented on dlio_benchmark/configs/workload/llama_70b_zero3.yaml:
```diff
@@ -0,0 +1,46 @@
+checkpoint:
+  checkpoint_folder: checkpoints/llama_70b
+  steps_between_checkpoints: 1
+  model_datatype: fp16
+  optimizer_datatype: fp32
```
I feel these should be in model. and you missed adding a new line at end of file here.

hariharan-devarajan replied:
These were not fixed.

hariharan-devarajan commented on dlio_benchmark/configs/workload/llama_7b.yaml:
```diff
@@ -0,0 +1,46 @@
+checkpoint:
+  checkpoint_folder: checkpoints/llama_7b
+  steps_between_checkpoints: 1
+  model_datatype: fp16
+  optimizer_datatype: fp32
```
I feel these should be in model. and you missed adding a new line at end of file here.

hariharan-devarajan replied:
This was not fixed.

hariharan-devarajan commented on dlio_benchmark/configs/workload/llama_7b_zero3.yaml:
```diff
@@ -0,0 +1,46 @@
+checkpoint:
+  checkpoint_folder: checkpoints/llama_7b_zero3
+  steps_between_checkpoints: 1
+  model_datatype: fp16
+  optimizer_datatype: fp32
```
I feel these should be in model. and you missed adding a new line at end of file here.

hariharan-devarajan replied:
This was not fixed.

hariharan-devarajan commented on dlio_benchmark/configs/workload/megatron_deepspeed.yaml:
Please keep the original Megatron deep speed. Rename it as LLNL version. And then put the ANL version in.

hariharan-devarajan commented on dlio_benchmark/main.py:
```diff
@@ -271,6 +274,7 @@ def _train(self, epoch):
                 self.stats.end_block(epoch, block, block_step)
                 self.stats.start_ckpt(epoch, block, overall_step)
                 self.checkpointing_mechanism.checkpoint(epoch, overall_step)
+                self.comm.barrier()
```
Why do we need an extra barrier here? I wanna avoid Barries if possible.

zhenghh04 replied:
This is for us to have meaningful throughput

hariharan-devarajan replied:
adding more barriers have performance implication where it stops matching real apps. Lets review this and see if we can avoid.

hariharan-devarajan commented on dlio_benchmark/main.py:
```diff
@@ -279,12 +283,14 @@ def _train(self, epoch):
             else:
                 block_step += 1
             overall_step += 1
+            self.overall_step_total += 1
```
Why do we need an extra barrier here? I wanna avoid Barries if possible.

hariharan-devarajan commented on dlio_benchmark/main.py:
```diff
@@ -279,12 +283,14 @@ def _train(self, epoch):
             self.stats.end_block(epoch, block, block_step)
             self.stats.start_ckpt(epoch, block, overall_step)
             self.checkpointing_mechanism.checkpoint(epoch, overall_step)
+            self.comm.barrier()
```
Why do we need an extra barrier here? I wanna avoid Barries if possible.

zhenghh04 replied:
This is for better calculating the throughput. In principle, because there is another barrier right after this. it is not adding additional cost here.

hariharan-devarajan replied:
Adding two Barries will increase cost at scale. I would not recommend it.

hariharan-devarajan reviewed (changes requested):
It mostly looks good. Apart from some code improvement suggestions.

zhenghh04 commented:
@hariharan-devarajan Please check again

hariharan-devarajan commented on dlio_benchmark/configs/workload/llama_8b_zero3.yaml:
```diff
@@ -0,0 +1,47 @@
+  checkpoint_folder: checkpoints/llama_7b_zero3
+  steps_between_checkpoints: 1
+  model_datatype: fp16
+  optimizer_datatype: fp32
```
add a new line at end.

hariharan-devarajan commented on dlio_benchmark/utils/statscounter.py:
```diff
@@ -375,6 +394,4 @@ def save_data(self):
-            logging.info(f"{utcnow()} outputs saved in RANKID_output.json")
-
-
+            logging.info(f"{utcnow()} outputs saved in RANKID_output.json")
```
missing new line at end of file.

hariharan-devarajan reviewed (changes requested):
Most of my comments were not addressed with no explaination. Lets discuss these if you want.

zhenghh04 commented:
@hariharan-devarajan ready for you to review again. 

I added two other features since last time we talked: 
* recovery support  (read)
* added checkpoint only (i.e., turning off train)

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -14,94 +14,319 @@
+        if self.args.hidden_size <= 0:
+            return 0
+        head_size = self.args.hidden_size//self.args.num_attention_heads
+        dim_kv = head_size * self.args.num_kv_heads        
```
missed this. `dim_kv`

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -14,94 +14,319 @@
+        mlp_4h_to_h = self.args.ffn_hidden_size*self.args.hidden_size
+        weight = self.args.hidden_size
+        lm_head = embedding
+        return embedding  + (input_norm + qkv + dense + layer_norm + mlp_h_to_4h + mlp_4h_to_h)*self.args.num_layers + weight + lm_head
```
what is `qkv` and `mlp_h_to_4h`? if `mlp_h_to_4h` is too big then at least add a line comment on what it is, when it is defined.

hariharan-devarajan commented on dlio_benchmark/checkpointing/base_checkpointing.py:
```diff
@@ -14,94 +14,319 @@
+
+    def get_layer_parameters(self, layer_index):
+        head_size = self.args.hidden_size//self.args.num_attention_heads
+        dim_kv = head_size * self.args.num_kv_heads
```
full form `dim_kv`

hariharan-devarajan reviewed (changes requested):
Almost there.

hariharan-devarajan reviewed (approved):
Looks good. Thank you for all the changes.
